sdk/src/main/java/software/amazon/awssdk/iot/AwsIotMqttConnectionBuilder.java

```diff
@@ -500,6 +505,74 @@ public AwsIotMqttConnectionBuilder withWebsocketCredentialsProvider(CredentialsP
         return this;
     }
 
+    /**
+     * A helper function to add parameters to the username in the withCustomAuthorizer function
+     */
+    private String addUsernameParameter(String inputString, String parameterValue, String parameterPreText, Boolean addedQuery) {
+        String return_string = inputString;
+        if (addedQuery == false) {
+            return_string += "?";
+        } else {
+            return_string += "&";
+        }
+
+        if (parameterValue.contains(parameterPreText)) {
+            return return_string + parameterValue;
+        } else {
+            return return_string + parameterPreText + parameterValue;
+        }
+    }
+
+    /**
+     * Configures the MQTT connection so it can use a custom authorizer.
+     * This function will modify the username, port, and TLS options.
+     *
+     * Note: All arguments are optional and can have "null" as valid input.
+     * See the description for each argument for information on what happens if null is passed.
+     * @param username The username to use with the custom authorizer. If null is passed, it will check to
+     *                 see if a username has already been set (via withUsername function). If no username is set then
+     *                 no username will be passed with the MQTT connection.
+     * @param authorizerName The name of the custom authorizer. If null is passed, then 'x-amz-customauthorizer-name'
+     *                       will not be added with the MQTT connection.
+     * @param signature The signature of the custom authorizer. If null is passed, then 'x-amz-customauthorizer-signature'
+     *                  will not be added with the MQTT connection.
+     * @param password The password to use with the custom authorizer. If null is passed, then no password will be set.
+     * @return
+     */
+    public AwsIotMqttConnectionBuilder withCustomAuthorizer(String username, String authorizerName, String signature, String password) {
+        isUsingCustomAuthorizer = true;
+        String username_string = "";
+        Boolean username_added_query = false;
+
+        if (username == null) {
+            if (config.getUsername() != null) {
+                username_string += config.getUsername();
+            }
+        } else {
+            username_string += username;
+        }
+
+        if (authorizerName != null) {
+            username_string = addUsernameParameter(username_string, authorizerName, "x-amz-customauthorizer-name=", username_added_query);
+            username_added_query = true;
+        }
+        if (signature != null) {
```

Comment:
Based on previous change, should this be checking if username_added_query is still false before proceeding since it may have been changed to true at this point?

Reply:
No, though this confused me initially as well which is why the previous change checked for it. We need to add it if the username is added or not so we can support having a username and a signature.

_________

Slightly off-topic note:

With a custom authorizer, you are not required to pass a username. It is possible you might have only a signature as the first parameter. It is also possible to have only a username, a username and signature, or nothing at all! All of the arguments you pass to a custom authorizer are optional, which is kinda strange but it's mentioned in the documentation and works. I'm guessing it is for different configurations and so it also supports default custom authroizers, which may just be an endpoint. It is rather strange though from my (limited) current understanding.

__________

I should change the name of the variable though, as it's really just for determining which character to use when adding it to the username string.